pkg/gardenlet/operation/botanist/dualstackmigration.go

```diff
@@ -154,17 +186,25 @@ func (b *Botanist) CheckPodCIDRsInNodes(ctx context.Context) error {
 func (b *Botanist) UpdateDualStackMigrationConditionIfNeeded(ctx context.Context) error {
 	shoot := b.Shoot.GetInfo()
 
+	nodeList := &corev1.NodeList{}
+	if b.ShootClientSet != nil {
+		if err := b.ShootClientSet.Client().List(ctx, nodeList); err != nil {
+			return fmt.Errorf("failed to list nodes during network stack migration: %w", err)
+		}
+	}
+
+	network, err := b.Shoot.Components.Extensions.Network.Get(ctx)
+	if err != nil {
+		return nil
+	}
+	networkReadyForDualStackMigration := len(network.Status.IPFamilies) == len(b.Shoot.GetInfo().Spec.Networking.IPFamilies)
+	updateFunction := b.DetermineUpdateFunction(networkReadyForDualStackMigration, nodeList)
+
 	constraint := v1beta1helper.GetCondition(shoot.Status.Constraints, gardencorev1beta1.ShootDualStackNodesMigrationReady)
-	if constraint == nil && len(shoot.Spec.Networking.IPFamilies) == 2 && shoot.Status.Networking != nil && len(shoot.Status.Networking.Nodes) == 1 {
-		if err := b.Shoot.UpdateInfoStatus(ctx, b.GardenClient, true, false, func(shoot *gardencorev1beta1.Shoot) error {
-			constraint := v1beta1helper.GetOrInitConditionWithClock(b.Clock, shoot.Status.Constraints, gardencorev1beta1.ShootDualStackNodesMigrationReady)
-			constraint = v1beta1helper.UpdatedConditionWithClock(b.Clock, constraint, gardencorev1beta1.ConditionProgressing, "DualStackMigration", "The shoot is migrating to dual-stack networking.")
-			shoot.Status.Constraints = v1beta1helper.MergeConditions(shoot.Status.Constraints, constraint)
-			return nil
-		}); err != nil {
+	if constraint == nil {
+		if err := b.Shoot.UpdateInfoStatus(ctx, b.GardenClient, true, false, updateFunction); err != nil {
 			return fmt.Errorf("failed updating %s constraint in shoot status: %w", gardencorev1beta1.ShootDualStackNodesMigrationReady, err)
 		}
 	}
-
 	return nil
```

Comment:
Let's reduce the scope of some variables and calculate them only when needed.

```suggestion

	if constraint := v1beta1helper.GetCondition(shoot.Status.Constraints, gardencorev1beta1.ShootDualStackNodesMigrationReady); constraint == nil {
		networkReadyForDualStackMigration := len(network.Status.IPFamilies) == len(b.Shoot.GetInfo().Spec.Networking.IPFamilies)
		updateFunction := b.DetermineUpdateFunction(networkReadyForDualStackMigration, nodeList)

		if err := b.Shoot.UpdateInfoStatus(ctx, b.GardenClient, true, false, updateFunction); err != nil {
			return fmt.Errorf("failed updating %s constraint in shoot status: %w", gardencorev1beta1.ShootDualStackNodesMigrationReady, err)
		}
	}

	return nil
```